contracts/Choice.sol

```diff
@@ -61,89 +70,89 @@ contract Choice {
+            cycles[currentCycleIndex].shares -= shares;
+        }
 
-        // todo: transfer tokens
         // todo: event
     }
 
     function vote(uint256 amount) external {
-        if (amount <= 0) revert ZeroAmount();
+        tokens += amount;  // TODO: send tokens in
+        updateCycle(amount);
 
         uint256 currentCycleIndex;
 
-        // cast the vote
-        if (cycles.length == 0) {
+        unchecked {  // updateCycle() will always add a cycle to cycles if none exists
+            currentCycleIndex = cycles.length - 1;
+
+            if (currentCycleIndex > 0) {
+                // There are no contributor fees in the cycle where the first contribution was made.
+                amount -= amount * contributorFee / 10000;
+            }
+        }
+
+        Vote[] storage votes = userVotes[msg.sender];
+
+        votes.push(
+            Vote({
+                cycleIndex: currentCycleIndex,
+                tokens: amount,
+                withdrawn: false
+            })
+        );
+
+        // todo: event
+    }
+
+    function updateCycle(uint256 amount) internal {
+        uint256 currentCycleNumber = ITopic(topicAddress).currentCycle();
+        uint256 length = cycles.length;
+
+        if (length == 0) { // Create the first cycle in the array using the first contribution.
             cycles.push(
-                CycleMetadata({
-                    cycle: ITopic(topicAddress).currentCycle(),
+                Cycle({
+                    cycle: currentCycleNumber,
                     shares: 0,
                     fees: 0,
                     hasVotes: true
                 })
             );
-            tokens = amount;
-        } else {
-            uint256 fee;
-            (currentCycleIndex, fee) = updateCycle(amount);
-            amount -= fee;
         }
+        else { // Not the first contribution.
 
-        // record voter metadata
-        VoteMetadata[] storage votes = userVotes[msg.sender];
-        uint256 length = votes.length;
-        if (length == 0 || votes[length - 1].cycleIndex != currentCycleIndex) {
-            votes.push(
-                VoteMetadata({
-                    cycleIndex: currentCycleIndex,
-                    tokens: amount,
-                    withdrawn: false
-                })
-            );
-        } else {
-            votes[length - 1].tokens += amount;
-        }
+            uint256 currentCycleIndex = length - 1;
 
-        // todo: transfer in tokens
-        // todo: event
-    }
+            Cycle storage currentCycle = cycles[currentCycleIndex];
 
-    function updateCycle(
-        uint256 amount
-    ) internal returns (uint256 cycleIndex, uint256 fee) {
-        uint256 currentCycle = ITopic(topicAddress).currentCycle();
-        cycleIndex = cycles.length - 1;
-        CycleMetadata memory lastCycle = cycles[cycleIndex];
+            uint256 fee;
 
-        if (lastCycle.cycle > 0) {
```

Comment:
This is the bug I was talking about in the main PR comment. 

When checking whether to charge a contributor fee, we need to check that the cycleIndex is 0, not that the cycleNumber is 0.